File: performance/hex-to-binary-performance/hex-to-bin-test.py
#!/usr/bin/env python3

import time

iterations = 100

# Read the hex file and convert it to a JPG file
def get_hex_data(hex_file_path):
   with open(hex_file_path, 'r') as hex_file:
      hex_data = hex_file.read()
      return hex_data

def hex_to_jpg(hex_data):
   # Convert hex to binary
   binary_data = bytes.fromhex(hex_data)
   return binary_data


def main():
   hex_data = get_hex_data('data/testdata.hex')

   start = time.time()

   # convert hex to binary
   # loop 100 times
   for i in range(iterations):
      #print('.', end='')
      binary_data = hex_to_jpg(hex_data)

   print()

   end = time.time()
   # format the time to microseconds
   print(f'Elapsed: {end - start:.6f} seconds')
   print(f'   Avg : {(end - start)/iterations:.6f} seconds')

   # Write binary data to a JPG file
   with open('data/python.dat', 'wb') as jpg_file:
      jpg_file.write(binary_data)


if __name__ == '__main__':
  main()

File: performance/hex-to-binary-performance/hex-to-bin-test.pl
#!/usr/bin/env perl
#
use strict;
use warnings;
use Time::HiRes qw(time);

use IO::File;

my $iterations = 100;

my $hex_file = 'data/testdata.hex';
$hex_file = 'data/testdata-100M.hex';
my $bin_file = 'data/perl.dat';

my $hex_fh = IO::File->new($hex_file, 'r') or die "Can't open $hex_file: $!";
my $bin_fh = IO::File->new($bin_file, 'w') or die "Can't open $bin_file: $!";
binmode $bin_fh;

my $hex_data;
while (my $line = $hex_fh->getline) {
	 chomp $line;
	 $line =~ s/^\s+//;
	 $line =~ s/\s+$//;
	 next if $line =~ /^#/;
	 $hex_data .= $line;
}

my $start_time = Time::HiRes::gettimeofday();

$| = 1;

my $bin_data;
for (my $i = 0; $i < $iterations; $i++) {
	#print '.';
	$bin_data = pack 'H*', $hex_data;
}

print "\n";

my $stop_time = Time::HiRes::gettimeofday();
printf("Elapsed: %.6f\n", $stop_time - $start_time);
printf("    Avg: %.6f\n", ($stop_time - $start_time)/$iterations);

$bin_fh->print($bin_data);



